openedx/core/djangoapps/xblock/README.rst

```diff
@@ -0,0 +1,94 @@
+   - If the user is unregistered, field data will be saved into the user's django session rather than the database.
+* To support sandboxed execution of XBlocks
+   - The runtime API provides a mechanism for calling XBlock handlers without session authentication (using a secure token in the URL), so that XBlocks can be run in sandboxed IFrames that cannot access the user's LMS cookies or interact with the LMS other than through the XBlock JavaScript API.
+
+Code Layout and Separation of Concerns
+--------------------------------------
+
+All of this code is found in this ``openedx.core.djangoapps.xblock`` folder:
+
+* `api.py <./api.py>`_ : **Python API** for this new XBlock runtime. Provides methods for loading and interacting with XBlocks given a user and a usage ID, such as: ``load_block()``, ``get_block_metadata()``, ``render_block_view()``, and ``get_handler_url()``
+
+  - Implementation differences in how Studio and the LMS implement these APIs are encapsulated in ``lms_impl.py`` and ``studio_impl.py``
+* `rest_api <./rest_api/>`_ : **REST API** for this new XBlock runtime. Provides `endpoints <./rest_api/urls.py>`_ for things like getting an XBlock's metadata, rendering an XBlock's view, calling an XBlock's handler, etc.
+  - An example URL is https://studio.example.com/api/xblock/v1/xblocks/lb:library1:html:introduction/view/student_view/
+* `runtime <./runtime/>`_ : **Blockstore-based XBlock runtime** which implements the complete XBlock runtime API and many Open edX-specific additions and runtime services.
+
+  - ``Scope.content`` and ``Scope.settings`` field data is handled by ``BlockstoreFieldData`` in `blockstore_field_data.py <./runtime/blockstore_field_data.py>`_ which knows how to read OLX files out of Blockstore given a ``BundleDefinitionLocator`` key. Since reading OLX out of Blockstore may be "slow", reads are cached per bundle (per bundle so that the cache is easily invalidated if Blockstore notifies us that the bundle has changed in any way). In the future, the LMS implementation of this runtime will likely load field data from a more optimized store, i.e. a more persistent version of the same cache.
+  - ``Scope.user_*`` and ``Scope.preferences`` field data is stored in the user's session (in Studio) or in new database tables (for the LMS) analogous to the existing runtime's ``StudentModule`` table.
+  - Each instance of the new runtime is designed to act as the runtime for multiple XBlocks at once (as long as they are requested by the same user), unlike the previous runtime which had to be instantiated separately for each XBlock. This is intended to be a memory and performance improvement, though it does require changes to some of the Open edX-specific runtime API extensions.
+  - Runtime code and services which can be shared among all instances of the runtime are encapsulated in a new ``XBlockRuntimeSystem`` singleton, again for memory and performance reasons.
```

Comment:
I'd like to understand this part better (I'll look through it in the code). As much as I am pro-performance, having this stuff live in a singleton opens us up to entire classes of bugs that we might not otherwise have to deal with. I had hoped that when we get rid of so much of the cruft that perf at this level wasn't much of an issue.

Reply:
To be frank, I wasn't aware of any specific memory or performance issues, but I'm sure this approach does use less memory and is more performant. That said, I went this way because:

1. It's trivial to convert back to the "one runtime instance per block" approach at any time if necessary, whereas once a lot of code has been written it's very difficult to change from "one runtime instance per block" to this "some or all parts of the runtime are shared by multiple blocks or even multiple requests".
1. Since we know there are going to be lots of requests for lots of XBlocks coming in all the time, constantly instantiating and destroying a runtime object with many attributes and services that are identical and often unused seems inefficient?
1. So far the runtime itself and the `XBlockRuntimeSystem` singleton essentially don't store any state, so I wasn't aware of those potential bugs.
1. The XBlock API seems clearly designed to have one runtime server multiple XBlock instances, which is why most of its methods take `block` as the first parameter.

There are a bunch of attributes on the old runtime that capa uses which assume the runtime is block-specific, but other than that I think it's possible to follow the "one runtime, many blocks" approach. All that said, I will defer to your experience with these things if you want me to change this.

Reply:
I agree that it's much easier to go from single to many than the other way, so let's roll with this for now. Thank you.